CloseIdleConnections of wrapped Transport RoundTrippers

aojea opened the pull request:
/kind bug
/kind cleanup

Fixes #100849

Add a function ` func CloseIdleConnectionsFor(transport http.RoundTripper) `  that allow to close the idle connections that belong to the transport, if the transport implements the `RoundTripperWrapper` interface.

Use this new function to close the idle connections on kubelet and avoid using a custom dialer that forces the kubelet to create multiple TCP connections , instead of reusing a single one.

```release-note
Don't use a custom dialer for the kubelet if is not rotating certificates, so we can reuse TCP connections and have only one between the apiserver and the kubelet.
If users experiment problems with stale connections  using HTTP1.1, they can force the previous behavior of the kubelet by setting the environment variable DISABLE_HTTP2.
```

aojea commented:
> E0908 16:39:20.689451    2358 server.go:293] "Failed to run kubelet" err="failed to run Kubelet: transport doesn't implement closeIdler: %!w(<nil>)"

this is not working, maybe I'm missing some wrapper :thinking:

fedebongio commented:
/assign @liggitt 
Jordan, giving it to you because people is asking you as part of https://github.com/kubernetes/kubernetes/issues/100849, hope you don't mind
/cc @caesarxuchao 
/triage accepted

caesarxuchao commented on staging/src/k8s.io/client-go/rest/connection_test.go:
```diff
@@ -162,3 +162,72 @@ func TestReconnectBrokenTCP(t *testing.T) {
+	if err != nil {
+		t.Fatalf("failed to create REST client: %v", err)
+	}
+	data, err := client.Get().AbsPath("/").DoRaw(context.TODO())
```
If the client is doing a WATCH instead of a GET, will the CloseIdleConnections on line 217 close the transport?

liggitt replied:
no, idle in this context doesn't mean "not sending/receiving data", it means "not handling a request and returned to the idle pool"

aojea replied:
if I understand the scenario correctly this tests does:
1. create an http1.1 connection ,  read, close the body and put it in the pool
2. break the tcp connection
3. try to reuse the connection, that should fail because TCP is broke

we solve it here removing the idle connections from the pool and forcing to create a new connection
(and I understand know Jordan's comment on the next commit :) )

aojea replied:
CloseIdleConnections() never closes connections that are not in the idle pool.
It turns out that once the request times out it is not returned to the idle pool
https://github.com/kubernetes/kubernetes/pull/104844#discussion_r707220875

liggitt replied:
> once the request times out it is not returned to the idle pool

Hmm... I'm skeptical that is true in all cases, since we saw persistently stuck kubelet -> apiserver connections that just timed out repeatedly ~forever (every few seconds for hours/days) back before http/2 was the default. If a request timing out was sufficient to keep the stuck connection from being returned to the pool, how would the kubelet have remained stuck using broken connections?

aojea replied:
just trying to find an explanation, can be this before golang 1.7? 1.7 introduced contexts and several important changes on the net/http library, I can see that client-go is using context cancellation for the requests too

liggitt replied:
it was still seen in the k8s 1.8-1.10 timeframe, which was on go 1.8-1.9

https://github.com/kubernetes/kubernetes/blob/release-1.10/hack/lib/golang.sh#L329

aojea replied:
is it possible that request weren't using timeout until this https://github.com/kubernetes/kubernetes/pull/51042 ?

liggitt replied:
timeout was set client-wide here in the k8s 1.10 timeframe:

https://github.com/kubernetes/kubernetes/blob/f170ef66340f6355d331ed90902574ff0532a20a/staging/src/k8s.io/client-go/rest/config.go#L210-L213

go says that behaves the same as if the context was canceled:

```
	// Timeout specifies a time limit for requests made by this
	// Client. The timeout includes connection time, any
	// redirects, and reading the response body. The timer remains
	// running after Get, Head, Post, or Do return and will
	// interrupt reading of the Response.Body.
	//
	// A Timeout of zero means no timeout.
	//
	// The Client cancels requests to the underlying Transport
	// as if the Request's Context ended.
	//
	// For compatibility, the Client will also use the deprecated
	// CancelRequest method on Transport if found. New
	// RoundTripper implementations should use the Request's Context
	// for cancellation instead of implementing CancelRequest.
	Timeout time.Duration
```

aojea replied:
yeah, but the client connections weren't timing out until 1.10, no?
let me try to see what happens if I disable the timeout  client side

aojea replied:
I think that the problem maybe related to pipelining and HOL blocking, I can reproduce it creating more than one connection
```go
func TestReconnectBrokenTCPAfterTimeout_HTTP1(t *testing.T) {
	const parallelRequests = 2
	mux := http.NewServeMux()
	mux.HandleFunc("/slow/", func(w http.ResponseWriter, r *http.Request) {
		// hold the connection
		t.Logf("Request received: %v", r.RemoteAddr)
		time.Sleep(10 * time.Second)
		fmt.Fprintf(w, "Slow Handler, %s", r.Proto)
	})
	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprintf(w, "Hello, %s", r.Proto)
	})
	ts := httptest.NewUnstartedServer(mux)
	ts.EnableHTTP2 = false
	ts.StartTLS()
	defer ts.Close()

	u, err := url.Parse(ts.URL)
	if err != nil {
		t.Fatalf("failed to parse URL from %q: %v", ts.URL, err)
	}
	lb := newLB(t, u.Host)
	defer lb.ln.Close()
	stopCh := make(chan struct{})
	go lb.serve(stopCh)
	transport, ok := ts.Client().Transport.(*http.Transport)
	if !ok {
		t.Fatalf("failed to assert *http.Transport")
	}
	config := &Config{
		Host:      "https://" + lb.ln.Addr().String(),
		Transport: utilnet.SetTransportDefaults(transport),
		Timeout:   3 * time.Second,
		// These fields are required to create a REST client.
		ContentConfig: ContentConfig{
			GroupVersion:         &schema.GroupVersion{},
			NegotiatedSerializer: &serializer.CodecFactory{},
		},
	}
	config.TLSClientConfig.NextProtos = []string{"http/1.1"}
	client, err := RESTClientFor(config)
	if err != nil {
		t.Fatalf("failed to create REST client: %v", err)
	}

	// 1. Create 10 parallel connections
	// 2. Once all of them connect close the TCP connections
	// 3. Wait until the connections timeout
	// 4. Try to connect again, if the connection is not reused it will succeed

	reqConnectCh := make(chan error)
	var once sync.Once
	trace := &httptrace.ClientTrace{
		GotConn: func(connInfo httptrace.GotConnInfo) {
			// t.Logf("Got Conn Reused %v WasIdle %v IdleTime %v", connInfo.Reused, connInfo.WasIdle, connInfo.IdleTime)
		},
		ConnectDone: func(network, addr string, err error) {
			//
		},
		WroteRequest: func(reqInfo httptrace.WroteRequestInfo) {
			// HTTP1.1 pipelining blocks until the first request is received
			once.Do(func() {
				t.Logf("First connection wrote request")
				reqConnectCh <- reqInfo.Err
				close(reqConnectCh)
			})
		},
	}
	ctx := httptrace.WithClientTrace(context.Background(), trace)
	// wait until all the requests times out
	wg := sync.WaitGroup{}
	reqTimeoutCh := make(chan struct{})
	go func() {
		for i := 0; i < parallelRequests; i++ {
			wg.Add(1)
			go func() {
				defer wg.Done()
				_, err := client.Get().AbsPath("/slow").DoRaw(ctx)
				t.Logf("Connection %d finished with err: %v", i, err)
			}()
		}
		wg.Wait()
		fmt.Println("Closing")
		close(reqTimeoutCh)
	}()

	select {
	case <-time.After(10 * time.Second):
		t.Fatalf("Test timed out waiting for first request to fail")
	case err := <-reqConnectCh:
		if err != nil {
			t.Fatalf("Not expected request to fail: %v", err)
		}
	}
	// Deliberately let the LB stop proxying traffic for the current
	// connection. This mimics a broken TCP connection that's not properly
	// closed.
	close(stopCh)

	stopCh = make(chan struct{})
	go lb.serve(stopCh)

	// Wait for the connection to time out
	select {
	case <-reqTimeoutCh:
		//for err := range errCh {
		//	if err == nil {
		//		t.Fatalf("Expected request to timeout")
		//	}
		//	t.Logf("Connection failed with error: %v", err)
		//}
	case <-time.After(10 * time.Second):
		t.Fatalf("Test timed out waiting for first request to fail")
	}

	// Close the idle connections
	utilnet.CloseIdleConnectionsFor(client.Client.Transport)

	// If the client didn't close the idle connections, the broken connection
	// would still be in the connection pool, the following request would
	// then reuse the broken connection instead of creating a new one, and
	// thus would fail.
	t.Logf("Creating new connection after network reconnected")
	data, err := client.Get().AbsPath("/").DoRaw(context.Background())
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	if string(data) != "Hello, HTTP/1.1" {
		t.Fatalf("unexpected response: %s", data)
	}
	dials := atomic.LoadInt32(&lb.dials)
	if dials != 2 {
		t.Fatalf("expected %d dials, got %d", 2, dials)
	}
}
```

aojea replied:
I can't reproduce it, the problem of the test was in the custom LB implementation that doesn't accept new connections

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/proxy/transport.go:
```diff
@@ -122,6 +122,16 @@ func (t *Transport) RoundTrip(req *http.Request) (*http.Response, error) {
 
 var _ = net.RoundTripperWrapper(&Transport{})
 
+func (rt *Transport) CloseIdleConnections() {
```
rather than making all round tripper layers implement this, would it make sense to implement a helper that calls WrappedRoundTripper() until it finds a round tripper that implements CloseIdleConnections()?

liggitt replied:
util/net#DialerFor is an example of what that could look like

liggitt commented on cmd/kubelet/app/server.go:
```diff
@@ -930,21 +929,15 @@ func buildKubeletClientConfig(ctx context.Context, s *options.KubeletServer, nod
-	d := connrotation.NewDialer((&net.Dialer{Timeout: 30 * time.Second, KeepAlive: 30 * time.Second}).DialContext)
-	clientConfig.Dial = d.DialContext
-	return d.CloseAll, nil
+	return clientConfig, tr.CloseIdleConnections, nil
```
I think there's still a hole if we only close idle connections

1. shared client is used by multiple goroutines
2. 2 connections are opened
3. both connections break (but don't close)
4. heartbeat fails on conn 1, while another goroutine is stuck waiting for its timeout to fire on conn 2
5. the heartbeat failure triggers calling CloseIdleConnections, which closes conn 1 but not conn 2
6. the heartbeat is reattempted, establishes a new working conn 3, and succeeds
7. the request over conn 2 finally times out, the request fails, and conn 2 is returned to the idle pool

That leaves us in a state where the heartbeat can be succeeding but we left broken connections in the idle pool

aojea replied:
is this about http2,  http1.1 or both? is the http2 hearbeat the one you mention?
in http2 you can not have 2 connections for the same destination.

liggitt replied:
http/1.1

liggitt replied:
the heartbeat is the node heartbeat goroutine... when that fails, it calls the "close all connections" function returned here

aojea replied:
> the request over conn 2 finally times out, the request fails, and conn 2 is returned to the idle pool

yeah, I can reproduce this behavior

aojea replied:
@liggitt http1.1 connections that times out are canceled and not returned to the idle pool, see test verifying it
```diff
diff --git a/staging/src/k8s.io/client-go/rest/connection_test.go b/staging/src/k8s.io/client-go/rest/connection_test.go
index c5027a0042b..7754c984f6b 100644
--- a/staging/src/k8s.io/client-go/rest/connection_test.go
+++ b/staging/src/k8s.io/client-go/rest/connection_test.go
@@ -23,6 +23,7 @@ import (
        "net"
        "net/http"
        "net/http/httptest"
+       "net/http/httptrace"
        "net/url"
        "os"
        "strconv"
@@ -43,6 +44,7 @@ type tcpLB struct {
 }
 
 func (lb *tcpLB) handleConnection(in net.Conn, stopCh chan struct{}) {
+       lb.t.Logf("Received connection from %v", in.RemoteAddr())
        out, err := net.Dial("tcp", lb.serverURL)
        if err != nil {
                lb.t.Log(err)
@@ -237,3 +239,113 @@ func TestReconnectBrokenTCP_HTTP1(t *testing.T) {
                t.Fatalf("expected %d dials, got %d", 2, dials)
        }
 }
+
+// 1. connect to https server with http1.1 using a TCP proxy making the connection to timeout
+// 2. the connection has keepalive enabled so it will be reused
+// 3. close the in-flight connection to force creating a new connection and avoid it goes to the idle pool once times out
+// 4. count that there are 2 connection on the LB but only one succeeds
+func TestReconnectBrokenTCPAfterTimeout_HTTP1(t *testing.T) {
+       ts := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+               // hold the connection
+               time.Sleep(2 * time.Second)
+               fmt.Fprintf(w, "Hello, %s", r.Proto)
+       }))
+       ts.EnableHTTP2 = false
+       ts.StartTLS()
+       defer ts.Close()
+
+       u, err := url.Parse(ts.URL)
+       if err != nil {
+               t.Fatalf("failed to parse URL from %q: %v", ts.URL, err)
+       }
+       lb := newLB(t, u.Host)
+       defer lb.ln.Close()
+       stopCh := make(chan struct{})
+       go lb.serve(stopCh)
+       transport, ok := ts.Client().Transport.(*http.Transport)
+       if !ok {
+               t.Fatalf("failed to assert *http.Transport")
+       }
+       config := &Config{
+               Host:      "https://" + lb.ln.Addr().String(),
+               Transport: utilnet.SetTransportDefaults(transport),
+               Timeout:   3 * time.Second,
+               // These fields are required to create a REST client.
+               ContentConfig: ContentConfig{
+                       GroupVersion:         &schema.GroupVersion{},
+                       NegotiatedSerializer: &serializer.CodecFactory{},
+               },
+       }
+       config.TLSClientConfig.NextProtos = []string{"http/1.1"}
+       client, err := RESTClientFor(config)
+       if err != nil {
+               t.Fatalf("failed to create REST client: %v", err)
+       }
+
+       reqErrCh := make(chan error, 1)
+       defer close(reqErrCh)
+       trace := &httptrace.ClientTrace{
+               GotConn: func(connInfo httptrace.GotConnInfo) {
+                       t.Logf("Got Conn Reused %v WasIdle %v IdleTime %v", connInfo.Reused, connInfo.WasIdle, connInfo.IdleTime)
+               },
+               ConnectDone: func(network, addr string, err error) {
+                       t.Logf("Connection done: %s %s", network, addr)
+               },
+               WroteRequest: func(reqInfo httptrace.WroteRequestInfo) {
+                       reqErrCh <- reqInfo.Err
+               },
+       }
+
+       // Create a connection and make it timeout closing the LB connection
+       // after it connects to the server
+       ctx := httptrace.WithClientTrace(context.Background(), trace)
+       errCh := make(chan error)
+       go func() {
+               _, err := client.Get().AbsPath("/").DoRaw(ctx)
+               errCh <- err
+       }()
+
+       select {
+       case err := <-reqErrCh:
+               if err != nil {
+                       t.Fatalf("Connection failed: %v", err)
+               }
+               // Deliberately let the LB stop proxying traffic for the current
+               // connection. This mimics a broken TCP connection that's not properly
+               // closed.
+               close(stopCh)
+
+               stopCh = make(chan struct{})
+               go lb.serve(stopCh)
+
+       case <-time.After(10 * time.Second):
+               t.Fatalf("Test timed out waiting for first request to fail")
+       }
+
+       // Wait for the connection to time out
+       select {
+       case err := <-errCh:
+               if err == nil {
+                       t.Fatalf("Connection succeeded but was expected to timeout")
+               }
+               t.Logf("Connection failed with error: %v", err)
+       case <-time.After(10 * time.Second):
+               t.Fatalf("Test timed out waiting for first request to fail")
+       }
+
+       // If the client didn't close the idle connections, the broken connection
+       // would still be in the connection pool, the following request would
+       // then reuse the broken connection instead of creating a new one, and
+       // thus would fail.
+       data, err := client.Get().AbsPath("/").DoRaw(ctx)
+       if err != nil {
+               t.Fatalf("unexpected err: %v", err)
+       }
+       if string(data) != "Hello, HTTP/1.1" {
+               t.Fatalf("unexpected response: %s", data)
+       }
+       dials := atomic.LoadInt32(&lb.dials)
+       if dials != 2 {
+               t.Fatalf("expected %d dials, got %d", 2, dials)
+       }
+}
```

aojea replied:
I'm going to add the test so we can avoid regressions on golang , the test seems stable.

```
stress ./rest.test -test.run "TestReconnect*"
5s: 0 runs so far, 0 failures
10s: 24 runs so far, 0 failures
15s: 24 runs so far, 0 failures
...
1m15s: 216 runs so far, 0 failures
1m20s: 216 runs so far, 0 failures
1m25s: 240 runs so far, 0 failures
1m30s: 264 runs so far, 0 failures
```
They are also really useful for troubleshooting these kind of problems

Commit:
CloseIdleConnections for wrapped Transport

It iterates over the wrapped transports until it finds one
that implements the CloseIdleConnections method and executes it.

add test for closeidle http1 connections

add test for http1.1 reconnect with inflight request

add test to reuse connection request

add test for request connect after timeout

add test for client-go request concurrency

aojea commented:
Jordan I think that some the roundetrippers is not implementing the WrapperRoundTrupper Interface

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/104844/pull-kubernetes-node-e2e-containerd/1436341443684732928/
```
W0910 15:23:08.097766   10968 plugins.go:132] WARNING: gce built-in cloud provider is now deprecated. The GCE provider is deprecated and will be removed in a future release. Please use https://github.com/kubernetes/cloud-provider-gcp
I0910 15:23:08.102740   10968 gce.go:933] Using existing Token Source &oauth2.reuseTokenSource{new:google.computeSource{account:"", scopes:[]string(nil)}, mu:sync.Mutex{state:0, sema:0x0}, t:(*oauth2.Token)(nil)}
I0910 15:23:08.105252   10968 server.go:579] "Successfully initialized cloud provider" cloudProvider="gce" cloudConfigFile=""
I0910 15:23:08.105388   10968 server.go:997] "Cloud provider determined current node" nodeName="e2e-3f269ba801-a7d53-master"
E0910 15:23:08.106637   10968 server.go:294] "Failed to run kubelet" err="failed to run Kubelet: closeAllConns must be a valid function other than nil"
````
found this related https://github.com/kubernetes/kubernetes/pull/50775, I'll keep digging.

In addition, is pending to resolve the problem of an http1 request that times out and its connection is returning to the idle pool ...

aojea commented:
this is working now

aojea commented on staging/src/k8s.io/client-go/rest/request_test.go:
```diff
@@ -2273,15 +2275,17 @@ func TestStream(t *testing.T) {
 		if err != nil {
 			t.Fatalf("failed to parse test URL: %v", err)
 		}
+		c = srv.Client()
```
otherwise fail with TLS enabled

aojea commented:
@liggitt @caesarxuchao moving the discussion to the main thread ...
I wasn't able to reproduce any issues with HTTP1.1 and using the new CloseIdleConnectionsFor() function, I spend some time thinking I have a repro for the issue you commented in https://github.com/kubernetes/kubernetes/pull/104844#discussion_r707843561 , but it was actually a problem with the "fake load balancer" used for testing.

I've added test cases asserting the behavior of client-go against timeouts and reconnections and others for verifying the reconnect behavior against TCP brokenness.
I think that this change is a big win for scalability https://github.com/kubernetes/kubernetes/issues/100849#issuecomment-919680681 and the risk seems constrained to http1.1 without certificates, what do you think?

aojea commented:
/retest
TestMetadataSharedInformerFactory

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/net/http.go:
```diff
@@ -238,6 +238,35 @@ func DialerFor(transport http.RoundTripper) (DialFunc, error) {
+		}
+		CloseIdleConnectionsFor(transport.WrappedRoundTripper())
+	default:
+		klog.Warningf("Transport doesn't implement CloseIdleConnections method")
```
prefer returning an error the caller can determine how to handle, rather than logging warnings inside a utility helper method

aojea replied:
then, I think that not returning nothing keeps the same behavior as the standard library

https://github.com/golang/go/blob/aa4e0f528e1e018e2847decb549cfc5ac07ecf20/src/net/http/client.go#L937-L951

and is consistent with the definition.

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/net/http.go:
```diff
@@ -238,6 +238,35 @@ func DialerFor(transport http.RoundTripper) (DialFunc, error) {
 
+// CloseIdleConnectionsFor close idles connections for the Transport.
+// If the Transport is wrapped it iterates over the wrapped round trippers
+// until it finds one that implemens the CloseIdleConnections method.
```
```suggestion
// until it finds one that implements the CloseIdleConnections method.
```

liggitt commented:
> I've added test cases asserting the behavior of client-go against timeouts and reconnections and others for verifying the reconnect behavior against TCP brokenness.
> I think that this change is a big win for scalability [#100849 (comment)](https://github.com/kubernetes/kubernetes/issues/100849#issuecomment-919680681) and the risk seems constrained to http1.1 without certificates, what do you think?

I'm not 100% convinced that all the previously observed issues are covered, but I don't have test cases to back that up :-/

If we merge this, and someone reports they are seeing issues again, what would we recommend they do to mitigate?

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/net/http.go:
```diff
@@ -238,6 +238,35 @@ func DialerFor(transport http.RoundTripper) (DialFunc, error) {
+	case *http.Transport:
+		transport.CloseIdleConnections()
+	case RoundTripperWrapper:
+		type closeIdler interface {
+			CloseIdleConnections()
+		}
```
yep

aojea commented:
> If we merge this, and someone reports they are seeing issues again, what would we recommend they do to mitigate?

if it happens again I've unfortunately introduced a regression that I have to fix :sweat_smile: , but this time we'll have more data to reproduce it. They should open a bug and assign it to me, so we can fix it forever

liggitt commented:
> if it happens again I've unfortunately introduced a regression that I have to fix 😅 , but this time we'll have more data to reproduce it. They should open a bug and assign it to me, so we can fix it forever

I think we need a better answer for them than "wait 1-2 months for a fix to land in a patch release"

aojea commented:
> I think we need a better answer for them than "wait 1-2 months for a fix to land in a patch release"

fair enough, let me work on that, 
/hold

mmiranda96 commented:
/triage accepted
/priority important-longterm

Commit:
don't use a custom dialer for the kubelet

Don't use a custom dialer for the kubelet if is not rotating
certificates, so we can reuse TCP connections because we don't need
a customer dialer.

Kubelet needs to be able to recover from stale http connections.
HTTP2 has a mechanism to detect broken connections by sending periodical pings.
HTTP1 only can have one persistent connection, and it will close all Idle connections
once the Kubelet heartbet fails. However, since there are many edge cases that we can't
control, users can still opt-in to the previous behavior for closing the connections by
setting the environment variable DISABLE_HTTP2.

aojea commented:
> I think we need a better answer for them than "wait 1-2 months for a fix to land in a patch release"

/hold cancel
@liggitt since the only case we have doubts is with HTTP1.1, and we already have an environment variable to DISABLE_HTTP2, users can force the previous behavior setting this variable

liggitt commented:
> @liggitt since the only case we have doubts is with HTTP1.1, and we already have an environment variable to DISABLE_HTTP2, users can force the previous behavior setting this variable

is that correct? `DISABLE_HTTP2` can force use of http/1.1, but if there's an issue with http/1.1, there's no way to force http/2, right?

liggitt commented:
The scenario I'm thinking of is a kubelet talking to a http/1.1 load balancer, where http/2 is not an option, regardless of the envvar. Load balancer-fronted API servers were exactly the scenario I remember being most problematic around the kubelet getting perma-stuck connections.

aojea commented:
> is that correct? `DISABLE_HTTP2` can force use of http/1.1, but if there's an issue with http/1.1, there's no way to force http/2, right?

HTTP2 is enabled by default, 

https://github.com/aojea/kubernetes/blob/90918a0a873f068b8be5d1e08f7eee8054d9c36e/staging/src/k8s.io/apimachinery/pkg/util/net/http.go#L125-L151

if they have http/1.1 has to be because they have some intermediate proxy, L4 loadbalancers are transparent.

> Load balancer-fronted API servers were exactly the scenario I remember being most problematic around the kubelet getting perma-stuck connections.

Yeah, this is the typical setup with nginx as reverse proxies, indeed is the more complex because there are a lot of points of failures, not only the network, users need to handle the certificates on the proxies too, ...

If something happens related to stuck connections it has to be http1.1 (or we have a new bug in http2). Since this env variable is documented, I just added it to recover previous behavior when closing the connections, in case that the new CloseIdleConnections() doesn't work.

liggitt commented:
> If something happens related to stuck connections it has to be http1.1 (or we have a new bug in http2). Since this env variable is documented, I just added it to recover previous behavior when closing the connections, in case that the new CloseIdleConnections() doesn't work (we don't have evidence yet)

I see... so the expectation would be:

1. kubelets using http/2 would use the http/2 ping timeout and this CloseIdleConnections mechanism
2. kubelets that automatically negotiated back to http/1.1 would use only this CloseIdleConnections mechanism
3. a kubelet could be forced to use http/1.1 and use the old tear-down-all-connections mechanism by explicitly setting DISABLE_HTTP2

That limits the exposure to regressions because of this new mechanism without an escape hatch to kubelets which:
1. don't use client certificate rotation
2. use http/2

That seems reasonable to me, though I'd like a sig-node reviewer to ack as well.

aojea commented:
> That seems reasonable to me, though I'd like a sig-node reviewer to ack as well.
absolutely

/assign @ehashman @SergeyKanzhelev

aojea commented:
> k8s.io/kubernetes/test/integration/job: TestNodeSelectorUpdate/feature_gate_disabled expand_less	

/test pull-kubernetes-integration

pacoxu commented:
/cc

pacoxu commented on cmd/kubelet/app/server.go:
```diff
@@ -933,9 +934,22 @@ func buildKubeletClientConfig(ctx context.Context, s *options.KubeletServer, nod
+	// Kubelet needs to be able to recover from stale http connections.
+	// HTTP2 has a mechanism to detect broken connections by sending periodical pings.
+	// HTTP1 only can have one persistent connection, and it will close all Idle connections
+	// once the Kubelet heartbet fails. However, since there are many edge cases that we can't
```
heartbeat

pacoxu commented on cmd/kubelet/app/server.go:
```diff
@@ -933,9 +934,22 @@ func buildKubeletClientConfig(ctx context.Context, s *options.KubeletServer, nod
+	// control, users can still opt-in to the previous behavior for closing the connections by
+	// setting the environment variable DISABLE_HTTP2.
+	var closeAllConns func()
+	if s := os.Getenv("DISABLE_HTTP2"); len(s) > 0 {
```
Should we add some log like below if http2 is disabled?
https://github.com/kubernetes/kubernetes/blob/b0bc8adbc2178e15872f9ef040355c51c45d04bb/staging/src/k8s.io/apimachinery/pkg/util/net/http.go#L135-L137

aojea replied:
I thought that will be logged anyway, because kubelet already calls `SetTransportDefaults`, however, now I don't know if this is the same codepath

```
pkg/kubelet/certificate/transport.go:   clientConfig.Transport = utilnet.SetTransportDefaults(&http.Transport{
```

aojea replied:
let me add the log with a more meaningful message about this behavior

aojea replied:
done

pacoxu replied:
/lgtm

pacoxu reviewed (approved):
/lgtm
One nit and a question if we should add an info log here.

pacoxu commented on cmd/kubelet/app/server.go:
```diff
@@ -933,9 +934,23 @@ func buildKubeletClientConfig(ctx context.Context, s *options.KubeletServer, nod
+	// setting the environment variable DISABLE_HTTP2.
+	var closeAllConns func()
+	if s := os.Getenv("DISABLE_HTTP2"); len(s) > 0 {
+		klog.Info("HTTP2 has been explicitly disabled, updating Kubelet client Dialer to forcefully close active connections on hearbeat failures")
```
```suggestion
		klog.InfoS("HTTP2 has been explicitly disabled, updating Kubelet client Dialer to forcefully close active connections on heartbeat failures")
```
For structured log.

aojea replied:
done

aojea commented:
/retest
> k8s.io/kubernetes/pkg/controller/volume/persistentvolume: TestControllerSync

ehashman commented:
/unassign

pacoxu commented:
/lgtm

aojea commented:
@liggitt for approval

aojea commented:
/assign @wojtek-t

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/util/net/http.go:
```diff
@@ -238,6 +238,29 @@ func DialerFor(transport http.RoundTripper) (DialFunc, error) {
+		CloseIdleConnectionsFor(transport.WrappedRoundTripper())
+	default:
+	}
+	return
```
nit: return is not needed

aojea replied:
:heavy_check_mark:

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/util/net/http.go:
```diff
@@ -238,6 +238,29 @@ func DialerFor(transport http.RoundTripper) (DialFunc, error) {
+		transport.CloseIdleConnections()
+	case RoundTripperWrapper:
+		CloseIdleConnectionsFor(transport.WrappedRoundTripper())
+	default:
```
nit: do we want to log some info/warning here?
In our usecases, this default is somewhat unexepected, so in case of issues it may help with some debugging, no?

aojea replied:
:heavy_check_mark:

wojtek-t commented on staging/src/k8s.io/client-go/rest/connection_test.go:
```diff
@@ -164,6 +164,189 @@ func TestReconnectBrokenTCP(t *testing.T) {
+		Host:      "https://" + lb.ln.Addr().String(),
+		Transport: utilnet.SetTransportDefaults(transport),
+		// large timeout, otherwise the broken connection will be cleaned by it
+		Timeout: 15 * time.Second,
```
nit: with our starved test environment, maybe we should go even higher?
maybe wait.ForeverTestTimeout ?

aojea replied:
done

wojtek-t commented on staging/src/k8s.io/client-go/rest/connection_test.go:
```diff
@@ -164,6 +164,189 @@ func TestReconnectBrokenTCP(t *testing.T) {
+		Host:      "https://" + lb.ln.Addr().String(),
+		Transport: utilnet.SetTransportDefaults(transport),
+		// Use something extraordinary large to not hit the timeout
+		Timeout: 60 * time.Second,
```
nit: unify with the comment I added above

aojea replied:
done

wojtek-t commented on staging/src/k8s.io/client-go/rest/connection_test.go:
```diff
@@ -164,6 +164,189 @@ func TestReconnectBrokenTCP(t *testing.T) {
+	// wait until it connect to the server
+	select {
+	case <-received:
+	case <-time.After(10 * time.Second):
```
nit: wait.ForeverTestTimeout

wojtek-t commented on staging/src/k8s.io/client-go/rest/connection_test.go:
```diff
@@ -164,6 +164,189 @@ func TestReconnectBrokenTCP(t *testing.T) {
+		if err == nil {
+			t.Fatal("Connection succeeded but was expected to timeout")
+		}
+	case <-time.After(10 * time.Second):
```
nit: wait.ForeverTestTimeout

wojtek-t commented on staging/src/k8s.io/client-go/rest/request_test.go:
```diff
@@ -2944,3 +2948,164 @@ func testRequestWithRetry(t *testing.T, key string, doFunc func(ctx context.Cont
+		enableHTTP2 bool
+	}{
+		{"HTTP1", false},
+		{"HTTP2", false},
```
should be true here, right?

aojea replied:
right

wojtek-t commented on staging/src/k8s.io/client-go/rest/request_test.go:
```diff
@@ -2944,3 +2948,164 @@ func testRequestWithRetry(t *testing.T, key string, doFunc func(ctx context.Cont
+		enableHTTP2 bool
+	}{
+		{"HTTP1", false},
+		{"HTTP2", false},
```
Same here

aojea replied:
interestingly, the http1 client shows a different behavior here than the http2, in http2 the connection is not closed after the timeout, I think that is because of

https://datatracker.ietf.org/doc/html/rfc7540#section-9.1
>    HTTP/2 connections are persistent.  For best performance, it is
   expected that clients will not close connections until it is
   determined that no further communication with a server is necessary
   (for example, when a user navigates away from a particular web page)
   or until the server closes the connection.

wojtek-t reviewed (commented):
> I see... so the expectation would be:

> 1. kubelets using http/2 would use the http/2 ping timeout and this CloseIdleConnections mechanism
> 2. kubelets that automatically negotiated back to http/1.1 would use only this CloseIdleConnections mechanism
> 3. a kubelet could be forced to use http/1.1 and use the old tear-down-all-connections mechanism by explicitly setting DISABLE_HTTP2

> That limits the exposure to regressions because of this new mechanism without an escape hatch to kubelets which:

> 1. don't use client certificate rotation
> 2. use http/2

> That seems reasonable to me, though I'd like a sig-node reviewer to ack as well.

That seems reasonable to me as well and it seems we have an ACK from sig-node reviewer.

I just added a couple nits to the PR.

wojtek-t commented:
/lgtm
/approve

JohnRusk commented:
@aojea Out of curiousity, how does the use of a custom dialer (before this PR) prevent re-use of connections?  I've been looking at the code in this PR because you said in another thread that it might explain the issue I've been seeing. (https://github.com/kubernetes/kubernetes/issues/65954#issuecomment-970047451)

aojea commented:
> @aojea Out of curiousity, how does the use of a custom dialer (before this PR) prevent re-use of connections? I've been looking at the code in this PR because you said in another thread that it might explain the issue I've been seeing. ([#65954 (comment)](https://github.com/kubernetes/kubernetes/issues/65954#issuecomment-970047451))

The client go constructor create a new transport if the dialer is customized

https://github.com/kubernetes/kubernetes/blob/8815a3119cc8bd82703f3fc5e5c715ec61dfc1ff/staging/src/k8s.io/client-go/transport/cache.go#L136

Before https://github.com/kubernetes/kubernetes/pull/105490 each group version created a new transport
